src/lib/rust.ts

```diff
@@ -328,130 +317,178 @@ const tryDownloadingBinary = async ({
       version,
-      installDependency: () => {
-        return installCargoDependency({
-          dependencyDirPath,
-          dependencyTmpDirPath,
-          name,
-          toolchain,
-          version,
-        });
-      },
+      toolchain,
     });
   }
 
-  await updateConfigsIfVersionChanged({
-    maybeFluenceConfig,
-    name,
-    version,
-    packageManager: "cargo",
-  });
+  const hasInstalledNotDefaultVersion = version !== versions.cargo[name];
+  const fluenceConfig = await initFluenceConfig();
+
+  const hasInstalledDefaultVersionButPreviouslyNotDefaultVersionWasUsed =
+    fluenceConfig?.[`${name}Version`] !== undefined &&
+    version === versions.cargo[name];
+
+  if (
+    fluenceConfig !== null &&
+    (hasInstalledNotDefaultVersion ||
+      hasInstalledDefaultVersionButPreviouslyNotDefaultVersionWasUsed)
+  ) {
+    fluenceConfig[`${name}Version`] = version;
+    await fluenceConfig.$commit();
+  }
 
   await addCountlyLog(`Using ${name}@${version} cargo dependency`);
   return dependencyDirPath;
-};
-
-type InstallAllDependenciesArg = {
-  maybeFluenceConfig: FluenceConfig | null;
-  force: boolean;
-};
+}
 
-export const installAllCargoDependencies = async ({
-  maybeFluenceConfig,
-  force,
-}: InstallAllDependenciesArg): Promise<void> => {
-  for (const [name, version] of Object.entries(
-    resolveCargoDependencies(maybeFluenceConfig),
-  )) {
+export async function ensureMarineAndMreplDependencies(): Promise<void> {
+  for (const [name, version] of await resolveMarineAndMreplDependencies()) {
     // Not installing dependencies in parallel
     // for cargo logs to be clearly readable
     // eslint-disable-next-line no-await-in-loop
-    await ensureCargoDependency({
-      nameAndVersion: `${name}@${version}`,
-      maybeFluenceConfig,
-      force,
-    });
+    await ensureMarineOrMreplDependency({ name, version });
   }
-};
+}
 
 type ResolveVersionArg = {
   name: string;
-  maybeVersion: string | undefined;
-  maybeFluenceConfig: FluenceConfig | null;
+  version: string | undefined;
 };
 
 async function resolveVersionToInstall({
   name,
-  maybeVersion,
-  maybeFluenceConfig,
+  version,
 }: ResolveVersionArg): Promise<string | undefined> {
-  if (typeof maybeVersion === "string") {
-    return (await isExactVersion(maybeVersion)) ? maybeVersion : undefined;
+  if (typeof version === "string") {
+    return (await isExactVersion(version)) ? version : undefined;
   }
 
-  const maybeRecommendedVersion = isFluenceCargoDependency(name)
-    ? versions.cargo[name]
+  const fluenceConfig = await initFluenceConfig();
+  return isMarineOrMrepl(name)
+    ? fluenceConfig?.[`${name}Version`] ?? versions.cargo[name]
     : undefined;
+}
+
+export async function resolveMarineAndMreplDependencies() {
+  const fluenceConfig = await initFluenceConfig();
+
+  return [
+    [
+      "marine",
+      fluenceConfig?.marineVersion === undefined
+        ? versions.cargo.marine
+        : fluenceConfig.marineVersion,
```

Comment:
```suggestion
      fluenceConfig?.marineVersion ?? versions.cargo.marine,
```

Reply:
The same thing below.
Not completely sure, but it looks like it's the usecase for null coalescing operator

Reply:
yeah, just had something more complex there before, didn't realize thing are actually simpler